File: apps/desktop/src/shared/api/effects.ts
import type { ApiEnvelope, PlayerState } from "./types";
import { isBoolean, isNumber, isRecord, isString } from "./ncmParserUtils";

export const EFFECTS_API_ROUTES = {
  setEq: { method: "POST", path: "/set_eq" },
  setEqType: { method: "POST", path: "/set_eq_type" },
  configureOptimizations: { method: "POST", path: "/configure_optimizations" },
  getCrossfeed: { method: "GET", path: "/crossfeed" },
  setCrossfeed: { method: "POST", path: "/set_crossfeed" },
  getSaturation: { method: "GET", path: "/saturation" },
  setSaturation: { method: "POST", path: "/set_saturation" },
  getDynamicLoudness: { method: "GET", path: "/dynamic_loudness" },
  setDynamicLoudness: { method: "POST", path: "/set_dynamic_loudness" },
  getNoiseShaperCurve: { method: "GET", path: "/noise_shaper_curve" },
  setNoiseShaperCurve: { method: "POST", path: "/set_noise_shaper_curve" },
  configureOutputBits: { method: "POST", path: "/configure_output_bits" }
} as const;

export type EffectsApiMethod = keyof typeof EFFECTS_API_ROUTES;
export type EffectsApiRoute = (typeof EFFECTS_API_ROUTES)[EffectsApiMethod];

export type EffectsRequestJson = (path: string, init?: RequestInit) => Promise<unknown>;
export type EffectsRequestEnvelope = (path: string, init?: RequestInit) => Promise<ApiEnvelope>;

export interface EffectsApiTransport {
  requestJson: EffectsRequestJson;
  requestEnvelope: EffectsRequestEnvelope;
}

export interface EffectsApiClient {
  setEq: (input: SetEqInput) => Promise<PlayerState>;
  setEqType: (input: SetEqTypeInput) => Promise<StatusMessageResponse>;
  configureOptimizations: (input: ConfigureOptimizationsInput) => Promise<PlayerState>;
  getCrossfeed: () => Promise<CrossfeedResponse>;
  setCrossfeed: (input: SetCrossfeedInput) => Promise<CrossfeedResponse>;
  getSaturation: () => Promise<SaturationResponse>;
  setSaturation: (input: SetSaturationInput) => Promise<SaturationResponse>;
  getDynamicLoudness: () => Promise<DynamicLoudnessResponse>;
  setDynamicLoudness: (input: SetDynamicLoudnessInput) => Promise<DynamicLoudnessResponse>;
  getNoiseShaperCurve: () => Promise<NoiseShaperResponse>;
  setNoiseShaperCurve: (input: SetNoiseShaperCurveInput) => Promise<NoiseShaperResponse>;
  configureOutputBits: (input: ConfigureOutputBitsInput) => Promise<StatusMessageResponse>;
}

export interface SetEqInput {
  bands?: Record<string, number>;
  enabled?: boolean;
}

export interface SetEqTypeInput {
  type: "IIR" | "FIR";
  fir_taps?: number;
}

export interface ConfigureOptimizationsInput {
  dither_enabled?: boolean;
  replaygain_enabled?: boolean;
}

export interface CrossfeedSettings {
  enabled: boolean;
  mix: number;
}

export interface SetCrossfeedInput {
  enabled?: boolean;
  mix?: number;
}

export interface SaturationSettings {
  enabled: boolean;
  drive: number;
  threshold: number;
  mix: number;
  input_gain_db: number;
  output_gain_db: number;
  highpass_mode: boolean;
  highpass_cutoff: number;
  sat_type?: string;
}

export interface SetSaturationInput {
  enabled?: boolean;
  drive?: number;
  threshold?: number;
  mix?: number;
  input_gain_db?: number;
  output_gain_db?: number;
  highpass_mode?: boolean;
  highpass_cutoff?: number;
}

export interface DynamicLoudnessSettings {
  enabled: boolean;
  strength: number;
  factor: number;
  band_gains: readonly number[];
}

export interface SetDynamicLoudnessInput {
  enabled?: boolean;
  strength?: number;
}

export type NoiseShaperCurve =
  | "Lipshitz5"
  | "FWeighted9"
  | "ModifiedE9"
  | "ImprovedE9"
  | "TpdfOnly";

export interface NoiseShaperSettings {
  curve: string;
  enabled: boolean;
  bits: number;
}

export interface SetNoiseShaperCurveInput {
  curve: NoiseShaperCurve;
}

export interface ConfigureOutputBitsInput {
  bits: 16 | 24 | 32;
}

export interface StatusMessageResponse {
  status: "success" | "error";
  message: string | null;
}

export interface CrossfeedResponse extends StatusMessageResponse {
  crossfeed: CrossfeedSettings;
}

export interface SaturationResponse extends StatusMessageResponse {
  saturation: SaturationSettings;
}

export interface DynamicLoudnessResponse extends StatusMessageResponse {
  dynamic_loudness: DynamicLoudnessSettings;
}

export interface NoiseShaperResponse extends StatusMessageResponse {
  noise_shaper: NoiseShaperSettings;
}

const parseStatus = (value: unknown): "success" | "error" => {
  if (value === "success" || value === "error") {
    return value;
  }
  throw new Error("Invalid effects response status");
};

const parseStatusMessage = (value: Record<string, unknown>): StatusMessageResponse => ({
  status: parseStatus(value.status),
  message: isString(value.message) ? value.message : null
});

const requireSuccess = (response: StatusMessageResponse, fallback: string) => {
  if (response.status === "error") {
    throw new Error(response.message ?? fallback);
  }
};

const parseCrossfeedSettings = (value: unknown): CrossfeedSettings | null => {
  if (!isRecord(value) || !isBoolean(value.enabled) || !isNumber(value.mix)) {
    return null;
  }
  return { enabled: value.enabled, mix: value.mix };
};

const parseSaturationSettings = (value: unknown): SaturationSettings | null => {
  if (
    !isRecord(value) ||
    !isBoolean(value.enabled) ||
    !isNumber(value.drive) ||
    !isNumber(value.threshold) ||
    !isNumber(value.mix) ||
    !isNumber(value.input_gain_db) ||
    !isNumber(value.output_gain_db) ||
    !isBoolean(value.highpass_mode) ||
    !isNumber(value.highpass_cutoff)
  ) {
    return null;
  }

  return {
    enabled: value.enabled,
    drive: value.drive,
    threshold: value.threshold,
    mix: value.mix,
    input_gain_db: value.input_gain_db,
    output_gain_db: value.output_gain_db,
    highpass_mode: value.highpass_mode,
    highpass_cutoff: value.highpass_cutoff,
    sat_type: isString(value.sat_type) ? value.sat_type : undefined
  };
};

const parseDynamicLoudnessSettings = (value: unknown): DynamicLoudnessSettings | null => {
  if (
    !isRecord(value) ||
    !isBoolean(value.enabled) ||
    !isNumber(value.strength) ||
    !isNumber(value.factor) ||
    !Array.isArray(value.band_gains) ||
    !value.band_gains.every(isNumber)
  ) {
    return null;
  }

  return {
    enabled: value.enabled,
    strength: value.strength,
    factor: value.factor,
    band_gains: value.band_gains
  };
};

const parseNoiseShaperSettings = (value: unknown): NoiseShaperSettings | null => {
  if (!isRecord(value) || !isString(value.curve) || !isBoolean(value.enabled) || !isNumber(value.bits)) {
    return null;
  }
  return { curve: value.curve, enabled: value.enabled, bits: value.bits };
};

const parsePayloadResponse = <T>(
  value: unknown,
  key: string,
  parser: (value: unknown) => T | null,
  errorMessage: string
): { response: StatusMessageResponse; payload: T } => {
  if (!isRecord(value)) {
    throw new Error(errorMessage);
  }
  const response = parseStatusMessage(value);
  requireSuccess(response, errorMessage);
  const payload = parser(value[key]);
  if (!payload) {
    throw new Error(errorMessage);
  }
  return { response, payload };
};

const parseStatusResponse = (value: unknown, errorMessage: string): StatusMessageResponse => {
  if (!isRecord(value)) {
    throw new Error(errorMessage);
  }
  const response = parseStatusMessage(value);
  requireSuccess(response, errorMessage);
  return response;
};

const requireState = (envelope: ApiEnvelope, fallback: string): PlayerState => {
  if (envelope.status === "error") {
    throw new Error(envelope.message ?? fallback);
  }
  if (!envelope.state) {
    throw new Error("State missing from effects response");
  }
  return envelope.state;
};

const postJson = (body: object): RequestInit => ({
  method: "POST",
  body: JSON.stringify(body)
});

export const setEq = async (transport: EffectsApiTransport, input: SetEqInput): Promise<PlayerState> => {
  const envelope = await transport.requestEnvelope(EFFECTS_API_ROUTES.setEq.path, postJson(input));
  return requireState(envelope, "Failed to update EQ");
};

export const setEqType = async (
  transport: EffectsApiTransport,
  input: SetEqTypeInput
): Promise<StatusMessageResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.setEqType.path, postJson(input));
  return parseStatusResponse(json, "Failed to set EQ type");
};

export const configureOptimizations = async (
  transport: EffectsApiTransport,
  input: ConfigureOptimizationsInput
): Promise<PlayerState> => {
  const envelope = await transport.requestEnvelope(
    EFFECTS_API_ROUTES.configureOptimizations.path,
    postJson(input)
  );
  return requireState(envelope, "Failed to configure optimizations");
};

export const getCrossfeed = async (transport: EffectsApiTransport): Promise<CrossfeedResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.getCrossfeed.path);
  const { response, payload } = parsePayloadResponse(
    json,
    "crossfeed",
    parseCrossfeedSettings,
    "Invalid crossfeed response"
  );
  return { ...response, crossfeed: payload };
};

export const setCrossfeed = async (
  transport: EffectsApiTransport,
  input: SetCrossfeedInput
): Promise<CrossfeedResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.setCrossfeed.path, postJson(input));
  const { response, payload } = parsePayloadResponse(
    json,
    "crossfeed",
    parseCrossfeedSettings,
    "Invalid crossfeed response"
  );
  return { ...response, crossfeed: payload };
};

export const getSaturation = async (transport: EffectsApiTransport): Promise<SaturationResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.getSaturation.path);
  const { response, payload } = parsePayloadResponse(
    json,
    "saturation",
    parseSaturationSettings,
    "Invalid saturation response"
  );
  return { ...response, saturation: payload };
};

export const setSaturation = async (
  transport: EffectsApiTransport,
  input: SetSaturationInput
): Promise<SaturationResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.setSaturation.path, postJson(input));
  const { response, payload } = parsePayloadResponse(
    json,
    "saturation",
    parseSaturationSettings,
    "Invalid saturation response"
  );
  return { ...response, saturation: payload };
};

export const getDynamicLoudness = async (
  transport: EffectsApiTransport
): Promise<DynamicLoudnessResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.getDynamicLoudness.path);
  const { response, payload } = parsePayloadResponse(
    json,
    "dynamic_loudness",
    parseDynamicLoudnessSettings,
    "Invalid dynamic loudness response"
  );
  return { ...response, dynamic_loudness: payload };
};

export const setDynamicLoudness = async (
  transport: EffectsApiTransport,
  input: SetDynamicLoudnessInput
): Promise<DynamicLoudnessResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.setDynamicLoudness.path, postJson(input));
  const { response, payload } = parsePayloadResponse(
    json,
    "dynamic_loudness",
    parseDynamicLoudnessSettings,
    "Invalid dynamic loudness response"
  );
  return { ...response, dynamic_loudness: payload };
};

export const getNoiseShaperCurve = async (
  transport: EffectsApiTransport
): Promise<NoiseShaperResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.getNoiseShaperCurve.path);
  const { response, payload } = parsePayloadResponse(
    json,
    "noise_shaper",
    parseNoiseShaperSettings,
    "Invalid noise shaper response"
  );
  return { ...response, noise_shaper: payload };
};

export const setNoiseShaperCurve = async (
  transport: EffectsApiTransport,
  input: SetNoiseShaperCurveInput
): Promise<NoiseShaperResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.setNoiseShaperCurve.path, postJson(input));
  const { response, payload } = parsePayloadResponse(
    json,
    "noise_shaper",
    parseNoiseShaperSettings,
    "Invalid noise shaper response"
  );
  return { ...response, noise_shaper: payload };
};

export const configureOutputBits = async (
  transport: EffectsApiTransport,
  input: ConfigureOutputBitsInput
): Promise<StatusMessageResponse> => {
  const json = await transport.requestJson(EFFECTS_API_ROUTES.configureOutputBits.path, postJson(input));
  return parseStatusResponse(json, "Failed to configure output bits");
};

export const createEffectsApiClient = (transport: EffectsApiTransport): EffectsApiClient => ({
  setEq: (input) => setEq(transport, input),
  setEqType: (input) => setEqType(transport, input),
  configureOptimizations: (input) => configureOptimizations(transport, input),
  getCrossfeed: () => getCrossfeed(transport),
  setCrossfeed: (input) => setCrossfeed(transport, input),
  getSaturation: () => getSaturation(transport),
  setSaturation: (input) => setSaturation(transport, input),
  getDynamicLoudness: () => getDynamicLoudness(transport),
  setDynamicLoudness: (input) => setDynamicLoudness(transport, input),
  getNoiseShaperCurve: () => getNoiseShaperCurve(transport),
  setNoiseShaperCurve: (input) => setNoiseShaperCurve(transport, input),
  configureOutputBits: (input) => configureOutputBits(transport, input)
});

File: src/server/effects.rs
use super::*;
use crate::config::normalize_eq_bands;
use actix_web::{web, HttpResponse};
use std::sync::Arc;

pub fn configure_routes(cfg: &mut web::ServiceConfig) {
    cfg.route("/set_eq", web::post().to(set_eq))
        .route("/set_eq_type", web::post().to(set_eq_type))
        .route(
            "/configure_optimizations",
            web::post().to(configure_optimizations),
        )
        .route("/crossfeed", web::get().to(get_crossfeed))
        .route("/set_crossfeed", web::post().to(set_crossfeed))
        .route("/saturation", web::get().to(get_saturation))
        .route("/set_saturation", web::post().to(set_saturation))
        .route("/dynamic_loudness", web::get().to(get_dynamic_loudness))
        .route(
            "/set_dynamic_loudness",
            web::post().to(set_dynamic_loudness),
        )
        .route("/noise_shaper_curve", web::get().to(get_noise_shaper_curve))
        .route(
            "/set_noise_shaper_curve",
            web::post().to(set_noise_shaper_curve),
        )
        .route(
            "/configure_output_bits",
            web::post().to(configure_output_bits),
        );
}

async fn set_eq(data: web::Data<Arc<AppState>>, body: web::Json<SetEqRequest>) -> HttpResponse {
    let normalized_bands = body.bands.as_ref().map(|bands| {
        normalize_eq_bands(bands.clone(), |unknown| {
            log::warn!("Unknown EQ band name: '{}'", unknown);
        })
    });

    if let Some(bands) = normalized_bands {
        let update = crate::config::EngineSettingsUpdate {
            eq_bands: Some(bands),
            ..crate::config::EngineSettingsUpdate::default()
        };
        if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update)
        {
            return crate::server::settings_handlers::audio_settings_error_response(error);
        }
    }

    if let Some(enabled) = body.enabled {
        let mut player = data.player.lock();
        if player.is_fir_eq_enabled() {
            if !enabled {
                player.disable_fir_eq();
            }
        } else {
            player.lockfree_eq_params.set_enabled(enabled);
        }
    }

    success_response("EQ updated")
}

async fn set_eq_type(
    data: web::Data<Arc<AppState>>,
    body: web::Json<SetEqTypeRequest>,
) -> HttpResponse {
    let eq_type_upper = body.eq_type.to_uppercase();
    if !matches!(eq_type_upper.as_str(), "IIR" | "FIR") {
        return bad_request_response(format!(
            "Unknown EQ type: '{}'. Supported types: IIR, FIR",
            body.eq_type
        ));
    }
    let update = crate::config::EngineSettingsUpdate {
        eq_type: Some(eq_type_upper.clone()),
        fir_taps: (eq_type_upper == "FIR").then_some(body.fir_taps.unwrap_or(1023)),
        ..crate::config::EngineSettingsUpdate::default()
    };
    match crate::server::settings_handlers::commit_settings_update(&data, update) {
        Ok(_) => HttpResponse::Ok().json(ApiResponse::success(format!(
            "EQ type set to {}",
            eq_type_upper
        ))),
        Err(error) => crate::server::settings_handlers::audio_settings_error_response(error),
    }
}

async fn configure_optimizations(
    data: web::Data<Arc<AppState>>,
    body: web::Json<ConfigureOptimizationsRequest>,
) -> HttpResponse {
    if let Some(dither_enabled) = body.dither_enabled {
        let update = crate::config::EngineSettingsUpdate {
            dither_enabled: Some(dither_enabled),
            ..crate::config::EngineSettingsUpdate::default()
        };
        if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update)
        {
            return crate::server::settings_handlers::audio_settings_error_response(error);
        }
    }
    success_response("Optimizations updated")
}

async fn set_crossfeed(
    data: web::Data<Arc<AppState>>,
    body: web::Json<SetCrossfeedRequest>,
) -> HttpResponse {
    let update = crate::config::EngineSettingsUpdate {
        crossfeed_enabled: body.enabled,
        crossfeed_mix: body.mix,
        ..crate::config::EngineSettingsUpdate::default()
    };
    if !update.is_empty() {
        if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update)
        {
            return crate::server::settings_handlers::audio_settings_error_response(error);
        }
    }
    success_response("Crossfeed updated")
}

async fn get_crossfeed(data: web::Data<Arc<AppState>>) -> HttpResponse {
    let player = data.player.lock();
    let settings = player.get_crossfeed_info();

    HttpResponse::Ok().json(serde_json::json!({
        "status": "success",
        "crossfeed": {
            "enabled": settings.enabled,
            "mix": settings.mix
        }
    }))
}

async fn set_saturation(
    data: web::Data<Arc<AppState>>,
    body: web::Json<SetSaturationRequest>,
) -> HttpResponse {
    let update = crate::config::EngineSettingsUpdate {
        saturation_enabled: body.enabled,
        saturation_drive: body.drive,
        saturation_mix: body.mix,
        ..crate::config::EngineSettingsUpdate::default()
    };
    if !update.is_empty() {
        if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update)
        {
            return crate::server::settings_handlers::audio_settings_error_response(error);
        }
    }

    {
        let player = data.player.lock();
        if let Some(threshold) = body.threshold {
            player.lockfree_saturation_params.set_threshold(threshold);
        }
        if let Some(input_gain_db) = body.input_gain_db {
            player
                .lockfree_saturation_params
                .set_input_gain(input_gain_db);
        }
        if let Some(output_gain_db) = body.output_gain_db {
            player
                .lockfree_saturation_params
                .set_output_gain(output_gain_db);
        }
        if let Some(highpass_mode) = body.highpass_mode {
            player
                .lockfree_saturation_params
                .set_highpass_mode(highpass_mode);
        }
        if let Some(highpass_cutoff) = body.highpass_cutoff {
            player
                .lockfree_saturation_params
                .set_highpass_cutoff(highpass_cutoff);
        }
    }
    success_response("Saturation updated")
}

async fn get_saturation(data: web::Data<Arc<AppState>>) -> HttpResponse {
    let player = data.player.lock();
    let settings = player.get_saturation_info();

    HttpResponse::Ok().json(serde_json::json!({
        "status": "success",
        "saturation": settings
    }))
}

async fn set_dynamic_loudness(
    data: web::Data<Arc<AppState>>,
    body: web::Json<SetDynamicLoudnessRequest>,
) -> HttpResponse {
    if let Some(strength) = body.strength {
        if !(0.0..=1.0).contains(&strength) {
            return bad_request_response("Strength must be between 0.0 and 1.0");
        }
    }
    let update = crate::config::EngineSettingsUpdate {
        dynamic_loudness_enabled: body.enabled,
        dynamic_loudness_strength: body.strength,
        ..crate::config::EngineSettingsUpdate::default()
    };
    if !update.is_empty() {
        if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update)
        {
            return crate::server::settings_handlers::audio_settings_error_response(error);
        }
    }
    success_response("Dynamic Loudness updated")
}

async fn get_dynamic_loudness(data: web::Data<Arc<AppState>>) -> HttpResponse {
    let player = data.player.lock();

    HttpResponse::Ok().json(serde_json::json!({
        "status": "success",
        "dynamic_loudness": {
            "enabled": player.is_dynamic_loudness_enabled(),
            "strength": player.get_dynamic_loudness_strength(),
            "factor": player.get_dynamic_loudness_factor(),
            "band_gains": player.get_dynamic_loudness_gains()
        }
    }))
}

async fn set_noise_shaper_curve(
    data: web::Data<Arc<AppState>>,
    body: web::Json<SetNoiseShaperCurveRequest>,
) -> HttpResponse {
    let curve = match body.curve.to_ascii_lowercase().as_str() {
        "lipshitz5" => crate::processor::NoiseShaperCurve::Lipshitz5,
        "fweighted9" => crate::processor::NoiseShaperCurve::FWeighted9,
        "modifiede9" => crate::processor::NoiseShaperCurve::ModifiedE9,
        "improvede9" => crate::processor::NoiseShaperCurve::ImprovedE9,
        "tpdfonly" => crate::processor::NoiseShaperCurve::TpdfOnly,
        _ => {
            return bad_request_response(format!(
                "Unknown noise shaper curve '{}'. Supported: Lipshitz5, FWeighted9, ModifiedE9, ImprovedE9, TpdfOnly",
                body.curve
            ));
        }
    };

    let update = crate::config::EngineSettingsUpdate {
        noise_shaper_curve: Some(format!("{:?}", curve)),
        ..crate::config::EngineSettingsUpdate::default()
    };
    if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update) {
        return crate::server::settings_handlers::audio_settings_error_response(error);
    }
    let (enabled, bits) = {
        let player = data.player.lock();
        (player.dither_enabled, player.get_output_bits())
    };
    HttpResponse::Ok().json(serde_json::json!({
        "status": "success",
        "message": format!("Noise shaper curve set to {:?}", curve),
        "noise_shaper": {
            "curve": format!("{:?}", curve),
            "enabled": enabled,
            "bits": bits
        }
    }))
}

async fn get_noise_shaper_curve(data: web::Data<Arc<AppState>>) -> HttpResponse {
    let player = data.player.lock();
    let curve = player.get_noise_shaper_curve();

    HttpResponse::Ok().json(serde_json::json!({
        "status": "success",
        "noise_shaper": {
            "curve": curve,
            "enabled": player.dither_enabled,
            "bits": player.get_output_bits()
        }
    }))
}

async fn configure_output_bits(
    data: web::Data<Arc<AppState>>,
    body: web::Json<SetOutputBitsRequest>,
) -> HttpResponse {
    if body.bits != 16 && body.bits != 24 && body.bits != 32 {
        return bad_request_response("Invalid bit depth. Supported: 16, 24, 32");
    }

    let update = crate::config::EngineSettingsUpdate {
        output_bits: Some(body.bits),
        ..crate::config::EngineSettingsUpdate::default()
    };
    if let Err(error) = crate::server::settings_handlers::commit_settings_update(&data, update) {
        return crate::server::settings_handlers::audio_settings_error_response(error);
    }
    HttpResponse::Ok().json(serde_json::json!({
        "status": "success",
        "message": format!("Output bit depth set to {} bits", body.bits)
    }))
}
